build/lib/test/test.js

```diff
@@ -0,0 +1,869 @@
+		return;
+	}
+
+	const actual = path.join(diffDir, 'actual.png');
+	await saveAppImage(details, actual);
+	try {
+		const diff = path.join(diffDir, 'diff.png');
+		await grabAppImage(details.platform, details.path.slice(0, -4) + '_diff.png', diff);
+	} catch (err) {
+		// ignore, diff image may not exist
+	}
+	return actual;
+}
+
+function generateJUnitPrefix(platform, target, deviceFamily) {
+	let prefix = platform;
+	if (target) {
+		prefix += '.' + target;
+	}
+	if (deviceFamily) {
+		prefix += '.' + deviceFamily;
+	}
+	return prefix;
+}
+
+/**
+ * Converts JSON results of unit tests into a JUnit test result XML formatted file.
+ *
+ * @param {Object} jsonResults JSON containing results of the unit test output
+ * @param {String} prefix prefix for test names to identify them uniquely
+ * @returns {Promise<void>}
+ */
+async function outputJUnitXML(jsonResults, prefix) {
+	// We need to go through the results and separate them out into suites!
+	const suites = {};
+	jsonResults.results.forEach(item => {
+		const s = suites[item.suite] || { tests: [], suite: item.suite, duration: 0, passes: 0, failures: 0, start: '' }; // suite name to group by
+		s.tests.unshift(item);
+		s.duration += item.duration;
+		if (item.state === 'failed') {
+			s.failures += 1;
+		} else if (item.state === 'passed') {
+			s.passes += 1;
+		}
+		suites[item.suite] = s;
+	});
+	const keys = Object.keys(suites);
+	const values = keys.map(v => suites[v]);
+	const template = await fs.readFile(JUNIT_TEMPLATE, 'utf8');
+	const r = ejs.render(template,  { suites: values, prefix: prefix });
+
+	// Write the JUnit XML to a file
+	const outFile = path.join(REPORT_DIR, `junit.${prefix}.xml`);
+	console.log(`JUnit test report written to ${outFile}`);
+	return fs.writeFile(outFile, r);
+}
+
+/**
+ * Remove all CI SDKs installed. Skip GA releases, and skip the passed in SDK path we just installed.
+ * @param  {String} sdkPath The SDK we just installed for testing. Keep this one in case next run can use it.
+ * @returns {Promise<void>}
+ */
+async function cleanNonGaSDKs(sdkPath) {
+	// FIXME Use fork since we're spawning off another node process!
+	const { stdout } = await exec(`node "${titanium}" sdk list -o json`);
+	const out = JSON.parse(stdout);
+	const installedSDKs = out.installed;
+	// Loop over the SDKs and remove any where the key doesn't end in GA, or the value isn't sdkPath
+	return Promise.all(Object.keys(installedSDKs).map(async item => {
+		const thisSDKPath = installedSDKs[item];
+		if (item.slice(-2) === 'GA') { // skip GA releases
+			return;
+		}
+		if (thisSDKPath === sdkPath) { // skip SDK we just installed
+			return;
+		}
+		console.log(`Removing ${thisSDKPath}`);
+		return fs.remove(thisSDKPath);
+	}));
+}
+
+/**
+ * @return {Promise<string>} path to Titanium SDK root dir
+ */
+async function sdkDir() {
+	try {
+		const { stdout, _stderr } = await exec(`node "${titanium}" config sdk.defaultInstallLocation -o json`);
```

Comment:
<!--
  0 failure: 
  1 warning:  build/lib/test/te...
  
  
  DangerID: danger-id-Danger;
  File: build/lib/test/test.js;
  Line: 695;
-->

- :warning: build/lib/test/test.js line 695 – '_stderr' is assigned a value but never used. (no-unused-vars)